### ISSUE TITLE
ERROR - invalid configuration: - 'pool_idle_time' must be at least 1 (after upgrade to 5.2.0)

### ISSUE THREAD

**Issue by thbar** [phusion/passenger#2020]
**Question 1: What is the problem?**

My app fails to start. In `/var/log/nginx/error.log`, I see:

```
# no errors before
2018/01/30 07:09:24 [info] 940#940: Using 32768KiB of shared memory for nchan in /etc/nginx/nginx.conf:73
ERROR: invalid configuration:
- 'pool_idle_time' must be at least 1
Raw configuration: {
   "config_manifest" : {
# SNIP (very long json)
```

The expected behaviour would be that the app would start (`pool_idle_time` is set to `0` in `/etc/nginx/passenger.conf`).

Here is the passenger config:

```
#Ansible managed

# see https://www.phusionpassenger.com/library/config/nginx/reference

passenger_root /usr/lib/ruby/vendor_ruby/phusion_passenger/locations.ini;
passenger_ruby /usr/bin/passenger_free_ruby;
# prefer always up
passenger_pool_idle_time 0; # using 1 here instead fixes the problem
```

The actual behaviour is a failure to start.

**Question 2: Passenger version and integration mode**:
- For example: open source 5.0.26 standalone; enterprise 5.0.21/nginx

Your answer: Passenger open-source for nginx, installed via APT (`5.2.0-1~xenial1`)

**Question 3: OS or Linux distro, platform (including version)**:
- For example: Debian 8, x86_64 or OS X 10.10 Yosemite, x86_64

Your answer: `Ubuntu 16.04.3 LTS`

**Question 4: Passenger installation method**:

Your answer:
[ ] RubyGems + Gemfile
[ ] RubyGems, no Gemfile
[x] Phusion APT repo
[ ] Phusion YUM repo
[ ] OS X Homebrew
[ ] source tarball
[ ] Other, please specify:

**Question 5: Your app's programming language (including any version managers) and framework (including versions)**:
- For example: Ruby 2.3.0, RVM, Rails 5.0.0; Node.js 4.0.0 with Express 4.13.4

Your answer: RVM, Ruby 2.3.5, Rails 5.0.3.

**Question 6: Are you using a PaaS and/or containerization? If so which one?**
- For example: Heroku, Amazon Container Services, Docker 1.9 with an image based on passenger-docker

Your answer: No

**Question 7: Anything else about your setup that we should know?**

Your answer:

This server is using the APT packages at `https://oss-binaries.phusionpassenger.com/apt/passenger` (`nginx-extras`, `passenger`).

It uses `unattended-upgrade` with automated reboot to get updates:

```yaml
unattended_automatic_reboot: true
unattended_remove_unused_dependencies: true
unattended_automatic_reboot_time: 07:00
unattended_origins_patterns:
  - 'origin=Ubuntu,archive=${distro_codename}-security,label=Ubuntu'
  - 'origin=packagecloud.io/phusion/passenger'
  - 'origin=apt.postgresql.org'
```

**Comment by antaflos** [comment]
Same problem here. The [docs still say that setting the pool idle time to 0 is valid](https://www.phusionpassenger.com/library/config/apache/reference/#passengerpoolidletime), which makes sense:

> Setting the value to 0 is recommended if you're on a non-shared host that's only running a few applications, each which must be available at all times.

I suspect this is a bug in Passenger rather than the docs.

**Comment by thbar** [comment]
Is there some recommendation available somewhere on how to "pin" Passenger to `5.1.x` until this is fixed? Given the number of currently opened issues (e.g. #2022 + the current ticket), it may be a good strategy.

Investigating how to use `/etc/apt/preferences` or how to "hold" the package on other servers until the  issues are ironed out.

**Comment by VvanGemert** [comment]
Got the same issue, this is a pretty nasty bug, because if you use default config and upgrade it will break passenger with this error. Even when setting passenger_pool_idle_time it will still break. I will not upgrade our production servers with 5.2 until this is resolved.

Our config:
- Phusion APT repo
- Ruby 2.5.0
- Passenger 5.1.12-1~trusty1 to 5.2.0-1~trusty1
- Ubuntu 14.04 LTS

Let me know if you need any more information related to this issue. Downgrading worked for me. I did it with the following command on my ubuntu version:

```apt-get install nginx-extras=1:1.12.2-8.5.1.12~trusty1 nginx-common=1:1.12.2-8.5.1.12~trusty1 passenger=1:5.1.12-1~trusty1```

**Comment by CamJN** [comment]
Sorry about this folks, we've identified the issue and created a patch here: https://github.com/phusion/passenger/tree/max_preloader_idle_time. It will be included in a follow-up release.

**Comment by thbar** [comment]
@CamJN thanks for the patch and upcoming release!

In general, I wonder:

* Are releases with minor number (e.g. 5.2 after 5.1) supposed to be backward compatible?
* Given the way your packagecloud apt publication works (I think it's a single package name, rather than e.g. `passenger-51`), is there a recommendation to "pin" to a certain version, maybe something that could go to the [apt install documentation](https://www.phusionpassenger.com/library/install/nginx/install/oss/xenial/)?

Thanks for any update you'll be able to provide.

**Comment by CamJN** [comment]
Releases which change the minor version number are not guaranteed to be backwards compatible, (though we generally try to keep the breakage to a minimum). For this release, even if this bug were not present, there are some deliberate breaking changes. You can read the [release blog post](https://blog.phusion.nl/2018/01/29/passenger-5-2-0/) for some details on the breaking changes. 

As for pinning an apt package, the method that I find easiest is to use `sudo apt-mark hold package_name` and then `sudo apt-mark unhold package_name` to revert to normal behaviour.

**Comment by thbar** [comment]
(Note that I realise that providing separate packages for each minor number is an overhead to the Passenger team - I'm not assuming that it's a transparent upgrade :smile:).

**Comment by thbar** [comment]
In case this helps someone else with the same issue, here is my fix for Xenial, leveraging the input from @VvanGemert and @CamJN (here in Ansible verbiage):

```yaml
- name: Install Passenger + Nginx
  # NOTE: use force=yes if you need to downgrade
  action: apt pkg={{item}} state=installed
  with_items:
    - nginx-extras=1:1.12.2-8.5.1.12~xenial1
    - nginx-common=1:1.12.2-8.5.1.12~xenial1
    - passenger=1:5.1.12-1~xenial1

# Quick adaptation from https://ansibledaily.com/idempotent-shell-command-in-ansible/
- name: Put packages on hold until 5.2 is more stable
  shell: >
         apt-mark showholds | grep -q passenger
         && echo -n HOLDED
         || apt-mark hold passenger nginx-extras nginx-common
  register: passenger_hold
  changed_when: passenger_hold.stdout != 'HOLDED'
```

**Comment by haslinger** [comment]
@CamJN you should definitely consider switching to semantic versioning

**Comment by sigv** [comment]
@CamJN I suppose that branch is not getting a PR where it can be tracked, so just going to ask outright for an ETA. As this blocks multiple of our services running on Passenger - when can we expect a point release to drop for this?

**Comment by sigv** [comment]
Would appreciate it even if it was something like _next month_ just to have clarity on this moving forward, you know. 😕

**Comment by OnixGH** [comment]
@sigv we don't have set dates for releases, the average cadence is about 3-4 weeks between releases, which can sway either way depending on issue analysis and severity.

This one is high prio, so we'll definitely do our utmost to have it in the next release. There's a few more things we're looking to include still, but ETA should be within the next couple of weeks.

**Comment by sigv** [comment]
@CamJN @OnixGH Not having any strict release schedules is totally understandable, but knowing the exact problem and the simple solution, why not expedite it through on a much shorter schedule? At least I believe point releases should do just that - quickly resolve issues, especially if they are regressions.

**Comment by OnixGH** [comment]
@sigv point releases don't save us any time, the release procedure is exactly the same for each release (it's a big procedure with building all sorts of packages and QA).
